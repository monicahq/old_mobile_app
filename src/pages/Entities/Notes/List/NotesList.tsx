import moment from 'moment';
import React, {PureComponent} from 'react';
import {ActivityIndicator, FlatList, Text, View} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';

import {EmptyActivity, Navbar, Touchable} from '@components';
import {I18n} from '@i18n';
import {INote} from '@models';
import {IPopAction} from '@navigator/NavigationService';
import {commonStyles, primaryColor} from '@theme';
import {styles} from './NotesList.styles';

interface INotesListProps {
  pop: IPopAction;
  getNotesByContact: () => void;
  navigateToNote: (noteId) => any;
  notes: INote[];
  isFetching: boolean;
}

export class NotesList extends PureComponent<INotesListProps, {}> {
  public componentWillMount() {
    this.props.getNotesByContact();
  }

  public keyExtractor = (item, index) => String(item.id);

  public renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  public renderItem = ({item, index}) => {
    const {notes, navigateToNote} = this.props;
    const note = notes[index];

    return (
      <Touchable onPress={navigateToNote(note.id)} style={styles.noteContainer}>
        <View style={{flexDirection: 'row'}}>
          <Text style={styles.date}>
            {moment(note.created_at).format('LL')}
          </Text>
          {note.is_favorited && (
            <Icon
              name="star"
              size={18}
              style={styles.favorite}
              color={primaryColor}
            />
          )}
        </View>
        <Text>{note.body}</Text>
      </Touchable>
    );
  };

  public render() {
    const {pop, notes, getNotesByContact, isFetching} = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title={I18n.t('notes:notes')} onBack={pop} />
        {isFetching || notes.length ? (
          <FlatList
            data={notes}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListFooterComponent={this.renderFooter}
            onEndReached={getNotesByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-notes.png')}
            title={I18n.t('notes:emptyTitle')}
            subtitle={I18n.t('notes:emptySubtitle')}
          />
        )}
      </View>
    );
  }
}
