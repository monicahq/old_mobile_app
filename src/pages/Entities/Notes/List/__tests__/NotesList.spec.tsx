import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {NotesList} from '../NotesList';

describe('Pages', () => {
  describe('Notes', () => {
    let pop;
    let getNotesByContact;
    let navigateToNote;

    beforeEach(() => {
      pop = jest.fn();
      getNotesByContact = jest.fn();
      navigateToNote = jest.fn();
    });

    it('should renders correctly without notes and fetching', () => {
      const notes = [];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={true}
          notes={notes}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly without notes and not fetching', () => {
      const notes = [];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={false}
          notes={notes}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly with notes and fetching', () => {
      const notes = [{body: 'My body', created_at: '1991-12-17'}];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={true}
          notes={notes as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
    });

    it('should renders correctly with notes and fetching', () => {
      const notes = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={false}
          notes={notes as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 1})))
      ).toMatchSnapshot();
    });

    it('should display the activity indicator in the footer if it is fetching', () => {
      const notes = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={true}
          notes={notes as any}
        />
      );
      expect(toJson(shallow(tree.instance().renderFooter()))).toMatchSnapshot();
    });

    it('should not display the activity indicator in the footer if it is not fetching', () => {
      const notes = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={false}
          notes={notes as any}
        />
      );
      expect(tree.instance().renderFooter()).toBeNull();
    });

    it('should have a keyExtractor', () => {
      const notes = [{id: 1}, {id: 2}];
      const tree = shallow(
        <NotesList
          pop={pop}
          navigateToNote={navigateToNote}
          getNotesByContact={getNotesByContact}
          isFetching={false}
          notes={notes as any}
        />
      );
      expect(tree.instance().keyExtractor(notes[0])).toBe('1');
      expect(tree.instance().keyExtractor(notes[1])).toBe('2');
    });
  });
});
