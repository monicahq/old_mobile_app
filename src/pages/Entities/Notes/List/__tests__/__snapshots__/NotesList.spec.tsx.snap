// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Notes should display the activity indicator in the footer if it is fetching 1`] = `
<ForwardRef(View)
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "marginVertical": 20,
      },
    ]
  }
>
  <Component
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    indeterminate={true}
    size="large"
    style={
      Object {
        "height": 36,
        "width": 36,
      }
    }
    styleAttr="Normal"
  />
</ForwardRef(View)>
`;

exports[`Pages Notes should renders correctly with notes and fetching 1`] = `
<ForwardRef(View)
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    rightAction={[MockFunction]}
    rightIcon="add"
    rightTitle="Add"
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "body": "My body",
          "created_at": "1991-12-17",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</ForwardRef(View)>
`;

exports[`Pages Notes should renders correctly with notes and fetching 2`] = `
<TouchableOpacity
  activeOpacity={0.2}
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <ForwardRef(View)
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <ForwardRef(Text)
      style={
        Object {
          "flex": 1,
          "fontSize": 12,
          "marginBottom": 8,
        }
      }
    >
      December 17, 1991
    </ForwardRef(Text)>
  </ForwardRef(View)>
  <ForwardRef(Text)>
    My body
  </ForwardRef(Text)>
</TouchableOpacity>
`;

exports[`Pages Notes should renders correctly with notes and fetching 3`] = `
<ForwardRef(View)
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    rightAction={[MockFunction]}
    rightIcon="add"
    rightTitle="Add"
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "body": "My body",
          "created_at": "1991-12-17",
        },
        Object {
          "body": "My body 2",
          "created_at": "1993-12-17",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</ForwardRef(View)>
`;

exports[`Pages Notes should renders correctly with notes and fetching 4`] = `
<TouchableOpacity
  activeOpacity={0.2}
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <ForwardRef(View)
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <ForwardRef(Text)
      style={
        Object {
          "flex": 1,
          "fontSize": 12,
          "marginBottom": 8,
        }
      }
    >
      December 17, 1991
    </ForwardRef(Text)>
  </ForwardRef(View)>
  <ForwardRef(Text)>
    My body
  </ForwardRef(Text)>
</TouchableOpacity>
`;

exports[`Pages Notes should renders correctly with notes and fetching 5`] = `
<TouchableOpacity
  activeOpacity={0.2}
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <ForwardRef(View)
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <ForwardRef(Text)
      style={
        Object {
          "flex": 1,
          "fontSize": 12,
          "marginBottom": 8,
        }
      }
    >
      December 17, 1993
    </ForwardRef(Text)>
  </ForwardRef(View)>
  <ForwardRef(Text)>
    My body 2
  </ForwardRef(Text)>
</TouchableOpacity>
`;

exports[`Pages Notes should renders correctly without notes and fetching 1`] = `
<ForwardRef(View)
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    rightAction={[MockFunction]}
    rightIcon="add"
    rightTitle="Add"
    title="Notes"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</ForwardRef(View)>
`;

exports[`Pages Notes should renders correctly without notes and not fetching 1`] = `
<ForwardRef(View)
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    rightAction={[MockFunction]}
    rightIcon="add"
    rightTitle="Add"
    title="Notes"
  />
  <EmptyActivity
    image={1}
    subtitle="Just write down what you have in mind."
    title="Notes are great to record precious details about your contacts."
  />
</ForwardRef(View)>
`;
